### PR TITLE
Update native code to upstream tag v2.8.0 (includes 2.7.0)

### PR DESCRIPTION
**Pull request by Kielek**
## What

Previous PR: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/4207

Sync native code from DD repository up to 2.8.0 tag.
All DD commits was manually reviewed, then I asked Codex/GPT 5.6 Sol Max to apply changes locally.
I hope, we will be able to sync further changes also. AI-assisted changes significantly speed up this process.

While reviewing, consider checking each commit separately, with corresponding DD changes.

I hope following notes will help during review

https://github.com/DataDog/dd-trace-dotnet/compare/v2.6.0...v2.7.0

https://github.com/DataDog/dd-trace-dotnet/commit/fb1bb827754f303e8f0fd1f700ddf037f48fc574 - profiler, skipped
**https://github.com/DataDog/dd-trace-dotnet/commit/4a7d43b991e003304e47b285cba7496710042d2d Add wildcard support to DD_TRACE_METHODS**
https://github.com/DataDog/dd-trace-dotnet/commit/6f0d718673f91b1d469464c15d667e95f072920a - Move loading dynamic library mechanism to shared code, skipped, as we do not have shared folder
https://github.com/DataDog/dd-trace-dotnet/commit/25c0025eb268ce5737e9b2adf9371aacc621277d - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/11bc8a067c5e867ddceefb2a4d51c9e3c8f1797c - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/543a79f9750104445d02b51e9aed444bdfc7b02b - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/5d88f2c33acc18bac27435a1ae0c559c1b0c8edc - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/1b1d09c2f334c65bbd6239f7ede62d4e1b9bef79 - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/23979f9bdee7c18525e41d25ea00007b25d882a4 - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/374e14e27356a8a8f7c68e1f31c1a4e14ace81d9 - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/443c791df73f42d5c2b28ff398c810a541add70b - already done in OTel .NET Auto
**https://github.com/DataDog/dd-trace-dotnet/commit/5df80e1810efd25d03ec2fb1172c9d5ac05e15f7 Automatically instrument methods decorated with [Trace] (Datadog.Trace.Annotations.TraceAttribute) Only ports the TypeReference parser refactor, TryRejitModule extraction, and native parser tests. The Datadog-specific TraceAttribute instrumentation is intentionally excluded.**
https://github.com/DataDog/dd-trace-dotnet/commit/39c57739e621d8a64b1dad916d1ab330d75bbed2 add support on osx - skipped as we already have it

https://github.com/DataDog/dd-trace-dotnet/compare/v2.7.0...v2.8.0

https://github.com/DataDog/dd-trace-dotnet/commit/66c61e1e7864999ff420deb6bbdffd608f2597d5 - profiler, skipped
**https://github.com/DataDog/dd-trace-dotnet/commit/a8fa5647d67e82bc83a8882298438a1007a0f10a Return CORPROF_E_PROFILER_CANCEL_ACTIVATION if Tracing is Disabled**
https://github.com/DataDog/dd-trace-dotnet/commit/8b13510cf5e31bd2fb646a702d101743b461e477 - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/fa58893a81893269eed4f5f7445dadd4c334a2ce - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/de477b69ece7a947f8ee2723250888023e9e2815 - profiler, skipped
https://github.com/DataDog/dd-trace-dotnet/commit/a040c9738b32e03e9b6d97db8312b07ae998c04d - profiler, skipped
**https://github.com/DataDog/dd-trace-dotnet/commit/a00c283881fd4baed1527e764dd921f28172a105 Add overload to create std::string from WCHAR* (+ size) **
**https://github.com/DataDog/dd-trace-dotnet/commit/e48254942320acf4d16048db650419547cb724b8 [DuckTyping] Two pass check to avoid creation of invalid types in memory - already done in #1717**
https://github.com/DataDog/dd-trace-dotnet/commit/effb44f385a983123ab47c055b2261bc58cac6a7 - profiler, skipped

## Tests

CI

## Checklist

- ~~[ ] `CHANGELOG.md` is updated.~~
- ~~[ ] Documentation is updated.~~
- [x] ~~New~~ features are covered by tests.

### CODE REVIEW

**Review by martincostello — APPROVED** [review]

**Review by nrcventura — APPROVED** [review]

**Review by rajkumar-rangaraj — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation.Native/integration.cpp [hunk 1]
```diff
@@ -37,11 +37,10 @@ AssemblyReference* AssemblyReference::GetFromCache(const WSTRING& str)
     return aref;
 }
 
-std::vector<IntegrationDefinition> GetIntegrationsFromTraceMethodsConfiguration(
-    const WSTRING& integration_assembly_name, const WSTRING& integration_type_name, const WSTRING& configuration_string)
+std::vector<IntegrationDefinition> GetIntegrationsFromTraceMethodsConfiguration(const TypeReference integration_type,
```
Comment (eftiquar) [inline src/OpenTelemetry.AutoInstrumentation.Native/integration.cpp:40]: Consider const TypeReference& integration_type. Non primtive consts are preferred to be const references.

File: src/OpenTelemetry.AutoInstrumentation.Native/string_utils.cpp [hunk 2]
```diff
@@ -25,24 +25,28 @@ std::string ToString(const uint64_t i)
...
 #ifdef _WIN32
-    if (wstr.empty())
+    if (nbChars == 0)
         return std::string();
 
     std::string tmpStr(tmp_buffer_size, 0);
-    int         size_needed =
-        WideCharToMultiByte(CP_UTF8, 0, &wstr[0], (int)wstr.size(), &tmpStr[0], tmp_buffer_size, NULL, NULL);
+    int size_needed = WideCharToMultiByte(CP_UTF8, 0, wstr, (int)nbChars, &tmpStr[0], tmp_buffer_size, NULL, NULL);
     if (size_needed < tmp_buffer_size)
```
Comment (eftiquar) [inline src/OpenTelemetry.AutoInstrumentation.Native/string_utils.cpp:39]: A pre-existing  issue has surfaced - WideCharToMultiByte's contract states that it returns zero and not size_needed if cbMultiByte  is not zero - here we are passing in tmp_buffer_size - ( which is 512). 
So any string > 256 wide char will be reported as empty string, because size_needed  will be set to zero and we will execute line 41 to return empty string.  We will need to call getlasterror if size_needed  is zero and look for ERROR_INSUFFICIENT_BUFFER and then recompute the needed size by passing in zero for cbMultiByte. This overhead is ok for rare occurrence of strings that do not fit in the default buffer size.

File: src/OpenTelemetry.AutoInstrumentation.Native/integration.cpp [hunk 3]
```diff
@@ -85,8 +83,17 @@ std::vector<IntegrationDefinition> GetIntegrationsFromTraceMethodsConfiguration(
 
             if (Logger::IsDebugEnabled())
             {
-                Logger::Debug("GetIntegrationsFromTraceMethodsConfiguration:  * Target: ", type_name, ".",
-                              method_definition, "(", signatureTypes.size(), ")");
+                if (method_definition == tracemethodintegration_wildcardmethodname)
```
Comment (eftiquar) [inline src/OpenTelemetry.AutoInstrumentation.Native/integration.cpp:86]: The wildcard semantics do not have any consumer in Otel. The wildcard filtering is used in Data Dog for specifying arbitrary methods for instrumenting. I would recommend we defer this until we bring that feature. 
Also the semantics imported from DD are confusing, inconsistent -  
1. Wildcard semantics are currently incoherent and name-based filtering is unreliable. * already denotes every eligible method, so configurations such as [*,Method] add no meaning and may submit the same MethodDef more than once. The grammar should either reject combining * with explicit methods or normalize the result into a unique MethodDef set.
2. If wildcard support is required, eligibility should be determined from CLR metadata before signature resolution and textual processing. Prefix checks for get_ and set_ are not reliable: an explicit-interface accessor can be named Namespace.IFoo.get_Value, and metadata does not require accessor classification to be inferred from that textual convention. MethodSemantics flags such as msGetter and msSetter provide the authoritative classification. Method attributes can similarly reject constructors, abstract methods, P/Invoke methods, and other unsupported categories.
3. The explicit-interface getter is an excellent concrete example, but the fundamental issue is broader: metadata semantics are authoritative; method-name conventions are merely incidental compiler output.
4. Wildcard support adds unnecessary complexity to existing machinery - The shared method-enumeration logic now has two modes: exact-name enumeration and whole-type enumeration.

File: src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.cpp [hunk 4]
```diff
@@ -606,6 +609,11 @@ HRESULT STDMETHODCALLTYPE CorProfiler::ModuleLoadFinished(ModuleID module_id, HR
         return S_OK;
     }
 
+    return TryRejitModule(module_id);
```
Comment (eftiquar) [inline src/OpenTelemetry.AutoInstrumentation.Native/cor_profiler.cpp:612]: The TryRejitModule method performs much more than ReJIT:

- Reads module metadata.
- Registers NGEN modules.
- Initializes CoreLib state.
- Injects/patches the loader.
- Redirects assembly references.
- Rewrites P/Invoke maps.
- Adds the module to module_ids_.
- Finally schedules ReJIT.
- Current gain: essentially none. There is still only one caller, and behavior is unchanged.

The name TryRejitModule is misleading—it sounds like a safe, focused ReJIT operation, but it processes and mutates nearly all module-load state, the caller must already hold module_ids_lock_.
I recommend omitting this extraction for now.

**Review by eftiquar — COMMENTED** [review]
My final recommendation 
1. Please keep the CORPROF_E_PROFILER_CANCEL_ACTIVATION improvement.
2. Retain the length-aware WCHAR conversion only after fixing the Windows two-pass conversion and error handling.
Omit wildcard support.
Omit the TryRejitModule extraction/refactor.
Omit supporting trace-method parser/signature refactors that exist only for wildcard.
Retain or add focused tests solely for cancellation behavior and string conversion, including null, empty, invalid UTF-16, and output larger than the temporary-buffer threshold.
That produces a small PR with two concrete correctness improvements and no unused feature machinery or unnecessary shared ReJIT refactoring.

File: src/OpenTelemetry.AutoInstrumentation.Native/string_utils.cpp (hunk 2 above)
Comment (eftiquar) [reply src/OpenTelemetry.AutoInstrumentation.Native/string_utils.cpp:39]: Also, the null-pointer check is required on both Windows and non-Windows paths.
The overload should begin with:

if (wstr == nullptr || nbChars == 0)
    {
        return {};
    }

**Comment by Kielek** [comment]
@eftiquar, I would apply your comments in case of the _standard_ PRs. This one is considered as a native code/duck type sync with the upstream. For now, my preference is

1. keep code changes as is.
2. made further syncs, maybe even to the lastly released code by DD?
3. Apply your comments in the separate PRs if decided to avoid any further syncs.

@zacharycmontoya, what is your/DD perspective on this PR and potential follow ups?

**Comment by eftiquar** [comment]
> @eftiquar, I would apply your comments in case of the _standard_ PRs. This one is considered as a native code/duck type sync with the upstream. For now, my preference is
> 
> 1. keep code changes as is.
> 2. made further syncs, maybe even to the lastly released code by DD?
> 3. Apply your comments in the separate PRs if decided to avoid any further syncs.
> 
> @zacharycmontoya, what is your/DD perspective on this PR and potential follow ups?

@Kielek, thank you for following up.

I appreciate the willingness to defer to follow-up PRs. Let me clarify my position on each concern:

**1. WideCharToMultiByte Bug (Must Fix):**
This is a pre-existing issue in OTel's `string_utils.cpp`, not introduced by this PR. Any string > 256 UTF-8 characters silently truncates to empty—a silent data loss bug affecting method names, type names, and configuration strings. Since we're reviewing string handling in this PR anyway, this is the right opportunity to fix it. I'd recommend we address this before merge.

**2. Wildcard Semantics & Code Quality (Defer with Recommendation):**
I'm perfectly comfortable accepting these as-is for this sync PR. However, they do introduce code bloat, hot-path filtering complexity, and correctness gaps (explicit-interface accessors). I'd strongly recommend DD consider adopting **CLR metadata flags** (`MethodSemantics`) instead of string-based filtering in their upstream implementation. This would benefit both DD and OTel, and make future syncs cleaner.

**3. TryRejitModule Extraction (Defer):**
No objection to deferring this to a follow-up PR.

**Summary:**
- Merge with **WideCharToMultiByte fix** (OTel's responsibility)
- Defer wildcard and TryRejitModule refinements to follow-up PRs (trusting the sync approach)
- Recommendation DD fix upstream with metadata-flags approach (mutual benefit for next sync)

Happy to discuss further or provide the multibyte fix if helpful.

File: src/OpenTelemetry.AutoInstrumentation.Native/string_utils.cpp (hunk 2 above)
Comment (zacharycmontoya) [reply src/OpenTelemetry.AutoInstrumentation.Native/string_utils.cpp:39]: Much later in v3.2.0 it looks like we provide a fix for the scenario where we have large strings. Can we defer the fix until that native update?

commit: https://github.com/DataDog/dd-trace-dotnet/commit/e0d9add131aa15e77624aaf334f6c31c89892aab

**Review by zacharycmontoya — APPROVED** [review]

**Comment by zacharycmontoya** [comment]
> @eftiquar, I would apply your comments in case of the _standard_ PRs. This one is considered as a native code/duck type sync with the upstream. For now, my preference is
> 
> 1. keep code changes as is.
> 2. made further syncs, maybe even to the lastly released code by DD?
> 3. Apply your comments in the separate PRs if decided to avoid any further syncs.
> 
> @zacharycmontoya, what is your/DD perspective on this PR and potential follow ups?

I think @Kielek's suggestion makes sense. I even noticed in one case that one of the bugs Eftiquar found was addressed in a follow-up DD change, so we can track several revisions that we should make and once we have sufficiently ported over the native bits from DD, we can begin to action on them individually

**Comment by eftiquar** [comment]
> > @eftiquar, I would apply your comments in case of the _standard_ PRs. This one is considered as a native code/duck type sync with the upstream. For now, my preference is
> > 
> > 1. keep code changes as is.
> > 2. made further syncs, maybe even to the lastly released code by DD?
> > 3. Apply your comments in the separate PRs if decided to avoid any further syncs.
> > 
> > @zacharycmontoya, what is your/DD perspective on this PR and potential follow ups?
> 
> I think @Kielek's suggestion makes sense. I even noticed in one case that one of the bugs Eftiquar found was addressed in a follow-up DD change, so we can track several revisions that we should make and once we have sufficiently ported over the native bits from DD, we can begin to action on them individually

Sure - if we plan to revisit this in next phase of sync up.
I am approving to allow the ducktype sync.
